packages/plugin-workspace-tools/sources/commands/foreach.ts

```diff
@@ -79,21 +80,29 @@ export default (clipanion: any, pluginConfiguration: PluginConfiguration) => cli
     async ({cwd, args, stdout, command, exclude, include, interlaced, parallel, topological, topologicalDev, all, verbose, jobs, ...env}: ForeachOptions) => {
       const configuration = await Configuration.find(cwd, pluginConfiguration);
       const {project, workspace} = await Project.find(configuration, cwd);
+      
+      if (!all && !workspace)
+        throw new WorkspaceRequiredError(cwd);
 
-      const rootWorkspace: Workspace | null = all ? project.topLevelWorkspace : workspace;
+      const rootWorkspace = all
+        ? project.topLevelWorkspace
+        : workspace;
 
-      const targetWorkspaces = rootWorkspace ? getWorkspaceChildrenRecursive(rootWorkspace, project) : [];
+      const candidates = getWorkspaceChildrenRecursive(rootWorkspace, project);
+      const workspaces = [];
 
-      // Prevents infinite loop in the case of configuring a script as such:
-      //     "lint": "yarn workspaces foreach --all lint"
-      const isRecursing = (workspace: Workspace) => command === process.env.npm_lifecycle_event && workspace.cwd === cwd;
-      let workspaces = targetWorkspaces.filter(workspace => workspace.manifest.scripts.has(command) && !isRecursing(workspace));
```

Comment:
I removed this line as I don't think they can recurse anymore, right? Since only the children will get traversed.

Reply:
They still can if `--all` is passed

Reply:
Actually yes, now they cannot - we exclude root workspace always

Reply:
This is a problem though. When the user passed `--all` the root workspace should be included as well and this check should be restored to prevent the recursion I think.

Reply:
Yep, that sounds reasonable, we can just start the `candidates` array with `rootWorkspace` as initial element 🤔

In this case I think the cwd check would be `workspace.cwd === rootWorkspace.cwd`, since the `cwd` might otherwise be deep within the folder.

Reply:
If `yarn workspaces foreach lint` was run inside the `lint` script, then `process.env.npm_lifecycle_event === command` and the cwd of package.json is `cwd`, not `rootWorkspace.cwd`. If process.env.npm_lifecycle_event !== command, then `yarn workspaces foreach lint` was run outside package.json, or the script name was different, e.g. `lint:all`.

Reply:
That doesn't hold when `cwd` is a subdirectory of the local workspace, though. If not rootWorkspace, maybe the workspace that gets returned by Project.find?

Reply:
Yep, sounds good to me